### PR TITLE
ATM2018Q2 - Hometask - M13

### PR DESCRIPTION
**Pull request by yauhen07**

### CODE REVIEW

File: src/main/java/core/CustomWebDriver.java [hunk 1]
```diff
@@ -1,17 +1,20 @@
...
+import org.openqa.selenium.*;
 
 import java.util.List;
 import java.util.Set;
 
-public class CustomWebDriver implements WebDriver {
-    protected WebDriver driver;
+public class CustomWebDriver implements WebDriver, JavascriptExecutor, TakesScreenshot {
+    private static final String RED_COLOR = "'#db3737'";
+    public WebDriver driver;
```
Comment (evgeny-tumashchick) [inline src/main/java/core/CustomWebDriver.java]: make private

File: src/main/java/core/CustomWebDriver.java [hunk 2]
```diff
@@ -1,17 +1,20 @@
...
+    public JavascriptExecutor javascriptExecutor;
+    public TakesScreenshot takesScreenshot;
```
Comment (evgeny-tumashchick) [inline src/main/java/core/CustomWebDriver.java]: remove JavascriptExecutor  and TakesScreenshot

File: src/main/java/core/CustomWebDriver.java [hunk 3]
```diff
@@ -32,7 +35,8 @@ public String getTitle() {
     }
 
     public WebElement findElement(By by) {
```
Comment (tumashchick) [inline src/main/java/core/CustomWebDriver.java]: public WebElement findElement(By by) {
        CustomLogger.info("Find element by" + by.toString());
        WebElement webElement = driver.findElement(by);
        executeScript("arguments[0].style.background=" + RED_COLOR, webElement);
        return webElement;
    }

File: src/main/java/core/CustomWebDriver.java [hunk 4]
```diff
@@ -68,4 +72,16 @@ public Navigation navigate() {
     public Options manage() {
         return driver.manage();
     }
+
+    public Object executeScript(String s, Object... objects) {
+        return javascriptExecutor.executeScript(s, objects);
```
Comment (evgeny-tumashchick) [inline src/main/java/core/CustomWebDriver.java]: return ((JavascriptExecutor) driver).executeScript(s, objects);

File: src/main/java/core/CustomWebDriver.java [hunk 5]
```diff
@@ -68,4 +72,16 @@ public Navigation navigate() {
...
+    }
+
+    public Object executeAsyncScript(String s, Object... objects) {
+        return javascriptExecutor.executeAsyncScript(s, objects);
```
Comment (evgeny-tumashchick) [inline src/main/java/core/CustomWebDriver.java]: return ((JavascriptExecutor) driver).executeAsyncScript(s, objects);

File: src/main/java/core/CustomWebDriver.java [hunk 6]
```diff
@@ -68,4 +72,16 @@ public Navigation navigate() {
...
+    }
+
+    public <X> X getScreenshotAs(OutputType<X> outputType) throws WebDriverException {
+        return takesScreenshot.getScreenshotAs(outputType);
```
Comment (evgeny-tumashchick) [inline src/main/java/core/CustomWebDriver.java]: return ((TakesScreenshot) driver).getScreenshotAs(outputType);

File: src/main/java/core/WebDriverSingleton.java [hunk 7]
```diff
@@ -1,5 +1,6 @@
 package core;
 
+import logging.CustomLogger;
```
Comment (tumashchick) [inline src/main/java/core/WebDriverSingleton.java]: Change WebDriver to Custom

public class WebDriverSingleton {

    private static CustomWebDriver instance;
    private static RunLocationCreator runLocationCreator;

    private WebDriverSingleton() {
    }

    public static void selectDriverType(RunLocationCreator locationCreator) {
        runLocationCreator = locationCreator;
    }

    public static CustomWebDriver getWebDriverInstance() {
        if (instance != null) {
            return instance;
        }
        return instance = init();
    }


    private static CustomWebDriver init() {
        if (null == runLocationCreator) {
            runLocationCreator = new LocalCreator();
        }
        WebDriver driver = runLocationCreator.getWebDriver();
        driver.manage().timeouts().pageLoadTimeout(20, TimeUnit.SECONDS);
        driver.manage().timeouts().implicitlyWait(20, TimeUnit.SECONDS);
        driver.manage().window().maximize();
        return new CustomWebDriver(driver);
    }

File: src/main/java/pageobject/AbstractPage.java [hunk 8]
```diff
@@ -1,5 +1,6 @@
 package pageobject;
 
+import core.CustomWebDriver;
```
Comment (evgeny-tumashchick) [inline src/main/java/pageobject/AbstractPage.java:3]: protected WebDriver driver; ->     protected CustomWebDriver driver;

File: src/main/java/pageobject/AbstractPage.java [hunk 9]
```diff
@@ -26,10 +27,10 @@ protected void waitForElementVisible(By locator) {
     }
 
     protected void highlightElement(By locator) {
```
Comment (tumashchick) [inline src/main/java/pageobject/AbstractPage.java]: protected void highlightElement(By locator) {
        driver.executeScript("arguments[0].style.background=" + RED_COLOR, driver.findElement(locator));
    }

    protected void unhighlightElement(By locator) {
        driver.executeScript("arguments[0].style.background=" + DEFAULT_COLOR, driver.findElement(locator));
    }

File: src/main/java/pageobject/LoginPage.java [hunk 10]
```diff
@@ -64,25 +57,22 @@ public HomePage inputCredAndLogin(User user) {
 
     public void inputUsernameHighlightAndScreenshotAction(String userName) {
```
Comment (evgeny-tumashchick) [inline src/main/java/pageobject/LoginPage.java:58]: change method names that do not take screenshot any more

File: src/main/java/pageobject/NewApplicationSecondFormPage.java [hunk 11]
```diff
@@ -12,22 +13,22 @@
...
     public void inputContactMobilePhone(String mobilePhone) {
-        System.out.println("Clear and populate Mobile Phone field");
+        CustomLogger.info("Clear and populate Mobile Phone field");
         Helpers.clearAndInput(driver, CONTACT_MOBILE_PHONE_INPUT_LOCATOR, mobilePhone);
     }
 
     public void selectBusinessLineOption(String businessLine) {
-        System.out.println("Select value from Business Line dropdown");
+        CustomLogger.info("Select value from Business Line dropdown");
         Helpers.selectValueFromDropdownAppForm(driver, BUSINESS_LINE_DROPDOWN_LOCATOR, BUSINESS_LINE_VALUE_LIST_LOCATOR, businessLine, "title");
```
Comment (evgeny-tumashchick) [inline src/main/java/pageobject/NewApplicationSecondFormPage.java]: remove driver as parameter, use getWebDriverInstance() in Helpers methods

File: src/main/java/util/ScreenshotsListener.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+import org.openqa.selenium.WebDriver;
+import org.testng.ITestResult;
+import org.testng.TestListenerAdapter;
+
+import java.io.File;
+import java.io.IOException;
+
+public class ScreenshotsListener extends TestListenerAdapter {
+    private static final String SCREENSHOTS_NAME_TPL = "screenshots/scr";
+    private WebDriver driver = WebDriverSingleton.getWebDriverInstance();
```
Comment (evgeny-tumashchick) [inline src/main/java/util/ScreenshotsListener.java]: no need in this

File: src/main/java/util/ScreenshotsListener.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+    @Override
+    public void onTestFailure(ITestResult result) {
+        if (!result.isSuccess()) {
+            File screenshot = ((TakesScreenshot) ((CustomWebDriver) driver).driver).getScreenshotAs(OutputType.FILE);
```
Comment (evgeny-tumashchick) [inline src/main/java/util/ScreenshotsListener.java]: File screenshot = WebDriverSingleton.getWebDriverInstance().getScreenshotAs(OutputType.FILE);

File: src/test/java/com/epam/atm/selenium/LoginPOTest.java [hunk 14]
```diff
@@ -7,7 +7,7 @@
 import pageobject.LoginPage;
 
 public class LoginPOTest extends BaseTest {
-    private static final String TITLE_HOME_PAGE = "Home | Salesforce";
+    private static final String TITLE_HOME_PAGE = "Home | Salesforce11";
```
Comment (evgeny-tumashchick) [inline src/test/java/com/epam/atm/selenium/LoginPOTest.java:10]: really changed to Salesforce11 ?  or u made this change to fail test?

**Review by evgeny-tumashchick — APPROVED** [review]

**Pull request merged**